File: frontend/src/app/controller/ProductCategoryController.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { ProductCategory } from "../entity/ProductCategory";
import { Observable } from "rxjs";
import { ProductCategorySearchDto } from "./search_dto/ProductCategorySearchDto";
import { Page } from "../dto/Page";

@Injectable({ providedIn: 'root' })
export class ProductCategoryController {

  constructor(private httpClient: HttpClient) { }

  save(productCategory: ProductCategory): Observable<ProductCategory> {
    return this.httpClient.post<ProductCategory>('http://localhost:8080/product-category/save', 
    productCategory);
  }

  update(productCategory: ProductCategory): Observable<ProductCategory> {
    return this.httpClient.post<ProductCategory>('http://localhost:8080/product-category/update', 
    productCategory);
  }

  delete(productCategory: ProductCategory): Observable<boolean> {
    return this.httpClient.post<boolean>('http://localhost:8080/product-category/delete', 
    productCategory)
  }

  search(productCategorySearchDto: ProductCategorySearchDto): Observable<Page<ProductCategory>> {
    return this.httpClient.post<Page<ProductCategory>>('http://localhost:8080/product-category/search', 
    productCategorySearchDto);
  }

}

File: ecommerce_api/src/main/java/personal/rajit/controller/ProductCategoryController.java
package personal.rajit.controller;

import lombok.AllArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import personal.rajit.controller.request_dto.ProductCategorySearchDto;
import personal.rajit.entity.ProductCategory;
import personal.rajit.service.ProductCategoryService;

@RestController
@RequestMapping("/product-category")
@AllArgsConstructor
public class ProductCategoryController {

    private final ProductCategoryService productCategoryService;

    @PostMapping("/save")
    public ResponseEntity<ProductCategory> save(@RequestBody ProductCategory productCategory) {
        return new ResponseEntity(productCategoryService.saveProductCategory(productCategory), HttpStatusCode.valueOf(200));
    }

    @PostMapping("/update")
    public ResponseEntity<ProductCategory> update(@RequestBody ProductCategory productCategory) {
        return new ResponseEntity(productCategoryService.updateProductCategory(productCategory), HttpStatusCode.valueOf(200));
    }

    @PostMapping("/delete")
    public ResponseEntity<String> delete(@RequestBody ProductCategory productCategory) {
        return new ResponseEntity(productCategoryService.deleteProductCategory(productCategory), HttpStatusCode.valueOf(200));
    }

    @PostMapping("/search")
    public ResponseEntity<Page<ProductCategory>> search(@RequestBody ProductCategorySearchDto productCategorySearchDto) {
        return new ResponseEntity(productCategoryService.searchProductCategory(productCategorySearchDto), HttpStatusCode.valueOf(200));
    }

}
